src/sources.rs

```diff
@@ -0,0 +1,346 @@
+    };
+
+    let mut interval = interval(Duration::from_secs(pull_interval.into()));
+    interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
+
+    loop {
+        interval.tick().await;
+        log::debug!("{}: Polling for exchange rate", source_label);
+
+        let raw_rate = match request_with_backoff(
+            || request_matcher(client.clone(), source.clone()),
+            |timeout: u64| {
+                log::warn!(
+                    "{}: Request not successful. Waiting for {} seconds until trying again",
+                    source_label,
+                    timeout
+                );
+                stats.increment_read_attempts(source_label);
+            },
+            INITIAL_RETRY_INTERVAL,
+            MAX_RETRIES,
+        )
+        .await
+        {
+            Some(i) => i,
+            None => {
+                log::error!("{}: Request failed. Retries exhausted", source_label);
+                stats.increment_read_attempts(source_label);
+                continue;
+            }
+        };
+        stats.reset_read_attempts(source_label);
+
+        if let Some(ref mut conn) = database_conn {
+            if let Err(e) = crate::database::write_read_rate(conn, raw_rate, source_label) {
+                stats.increment_failed_database_updates();
+                log::error!(
+                    "{}: Unable to INSERT new reading: {}, due to: {}",
+                    source_label,
+                    raw_rate,
+                    e
+                )
+            };
+        }
+        stats.update_read_rate(raw_rate, &source_label);
+
+        let rate = match BigRational::from_float(raw_rate) {
+            Some(r) => r.recip(), // Get the inverse value, to change units from EUR/CCD to CCD/EUR
+            None => {
+                log::error!("{}: Unable to convert rate to rational: {}", source_label, raw_rate);
+                continue;
+            }
+        };
+        log::info!("{}: New exchange rate polled: {}/{}", source_label, rate.numer(), rate.denom());
+        {
+            let mut rates = rate_history_mutex.lock().unwrap();
+            rates.push_back(rate);
+            if rates.len() > max_rates_saved {
+                rates.pop_front();
+            }
+        }; // drop lock
+    }
+}
+
+#[derive(SerdeDeserialize)]
+struct CoinMarketCapResponsePrice {
+    // TODO: add other fields like volume and change
+    price: f64,
+}
+
+#[derive(SerdeDeserialize)]
+struct CoinMarketCapResponseEur {
+    #[serde(rename = "EUR")]
+    eur: CoinMarketCapResponsePrice,
+}
+
+#[derive(SerdeDeserialize)]
+struct CoinMarketCapResponseInfo {
+    // TODO: add other fields about CCD
+    quote: CoinMarketCapResponseEur,
+}
+
+#[derive(SerdeDeserialize)]
+struct CoinMarketCapResponseData {
+    #[serde(rename = "CCD")]
+    ccd: Vec<CoinMarketCapResponseInfo>,
+}
+
+#[derive(SerdeDeserialize)]
+pub struct CoinMarketCapResponse {
+    // TODO: add status structure
```

Comment:
What is in the status? Does it contain useful information which would cause us to reject readings? If that's the case we should do it now.